src/app/modules/bounce/bounce.service.ts

```diff
@@ -110,26 +116,68 @@ const logCriticalBounce = (
   bounceDoc: IBounceSchema,
   submissionId: string | undefined,
   bounceInfo: IBounceNotification['bounce'] | undefined,
+  autoEmailRecipients: string[],
 ): void => {
-  if (bounceDoc.isCriticalBounce()) {
+  logger.warn({
+    message: 'Critical bounce',
+    meta: {
+      action: 'updateBounces',
+      hasAutoEmailed: bounceDoc.hasEmailed,
+      formId: String(bounceDoc.formId),
+      submissionId: submissionId,
+      recipients: bounceDoc.getEmails(),
+      autoEmailRecipients,
+      // We know for sure that critical bounces can only happen because of bounce
+      // notifications, so we don't expect this to be undefined
+      bounceInfo: bounceInfo,
+    },
+  })
+}
+
+const computeValidEmails = (
+  populatedForm: IPopulatedForm,
+  bounceDoc: IBounceSchema,
+): string[] => {
+  const collabEmails = populatedForm.permissionList
+    ? populatedForm.permissionList.map((collab) => collab.email)
+    : []
+  const possibleEmails = collabEmails.concat(populatedForm.admin.email)
+  return difference(possibleEmails, bounceDoc.getEmails())
+}
+
+const handleCriticalBounce = async (
+  bounceDoc: IBounceSchema,
+  submissionId: string | undefined,
+  bounceInfo: IBounceNotification['bounce'] | undefined,
+): Promise<void> => {
+  const form = await Form.getFullFormById(bounceDoc.formId)
+  if (!form) {
     logger.warn({
-      message: 'Critical bounce',
+      message: 'Unable to retrieve form',
       meta: {
         action: 'updateBounces',
-        hasAlarmed: bounceDoc.hasAlarmed,
-        formId: String(bounceDoc.formId),
-        submissionId: submissionId,
-        recipients: bounceDoc.bounces.map((emailInfo) => emailInfo.email),
-        // We know for sure that critical bounces can only happen because of bounce
-        // notifications, so we don't expect this to be undefined
-        bounceInfo: bounceInfo,
+        formId: bounceDoc.formId,
       },
     })
-    // TODO (private #31): autoemail and set hasAlarmed to true. Currently
-    // hasAlarmed is a dangling key.
-    bounceDoc.hasAlarmed = true
-    // TODO (private #31): convert bounceType to enum.
+    return
+  }
+  // TODO (private #30): enable form deactivation
+  // if (bounceInfo?.bounceType === BounceType.Permanent) {
+  //   await form.deactivate()
+  // }
+  const emailRecipients = computeValidEmails(form, bounceDoc)
+  if (emailRecipients.length > 0 && !bounceDoc.hasEmailed) {
```

Comment:
move up the `!bounceDoc.hasEmailed` check to save work on computing valid emails unnecessarily.